### ISSUE TITLE
Copilot is about to start charging.

### ISSUE THREAD

**Issue by CharlesChiuGit** [zbirenbaum/copilot.lua#38]
As the letter many of copilot users received today, maybe it's time to look for a open-source AI coding assistant?
I found [this cool project](https://github.com/CodedotAl/gpt-code-clippy) in github, maybe there's a way to transform it into a Neovim plugin? This is more like a plugin idea, no need to take it seriously if you're not interested in making one.

Hope this issue won't bother you too much, feel free to close this issue if you think it's irrelevant.

Best!

**Comment by zbirenbaum** [comment]
Believe it or not I am actually way ahead of you. I've been looking at different options for a while that I could add support or create a new plugin for.

There are a few issues I have run into though. The model that you linked doesn't really have a full interface, and to take advantage of the neovim features I do to keep things simple, it would require me to implement a LSP server. I looked just now and it seems there is a vscode extension I missed before that was built on it so that may be of help. I could make a simpler rpc program, but it would require a lot of changes to this code base, and would be more difficult to debug. Plus, adding new features would be much more doable with the existing LSP infrastructure in neovim.

That said, I was actually planning to write my own LSP for the OpenAI codex engines which power copilot, as it would make someone with an api key for either eligible to use this plugin. I thought there would be a few months at the very least before copilot announced the paid model, so this caught me off guard and moves up my time frame quite a bit. 

The model you linked is the most promising truly free option I managed to run across as well, but I've heard gpt2 is quite a bit behind gpt3. Given that, I would **highly** recommend you apply for an OpenAI codex key in the meantime, as that is likely going to be your best bet for getting comparable completions to those copilot provides once I've completed a LSP server for it.

I'll go on record and state that I'll do my best to have a LSP implementation for an alternative by August 22nd.

**Comment by zbirenbaum** [comment]
I'm gonna pin this so that people won't miss it and can provide input on potential sources or my plans to write a more generalized LSP sever

**Comment by CharlesChiuGit** [comment]
Wow! That's super great news to know! Thanks for your study and effort! 
And don't push yourself too hard, take your time!🤣

**Comment by marcelarie** [comment]
Got my fresh Open AI key :) How is this going??

**Comment by zbirenbaum** [comment]
I have a functioning prototype but it gets rate limited really fast, as I found out upon completing it that open ai limits the free usage to 20 requests per minute. Trying to find a way around that and to integrate open source models :(

**Comment by CharlesChiuGit** [comment]
Thanks again for your effort!

**Comment by marcelarie** [comment]
thanks!

**Comment by marcelarie** [comment]
Yes I would love to see this happening. @zbirenbaum if you need any help please post it here, maybe I can't, but someone else can.

**Comment by nyngwang** [comment]
@rj1 Your link is incorrect.

**Comment by McPatate** [comment]
Hi @zbirenbaum, I've started writing a [plugin](https://github.com/huggingface/hfcc.nvim) when the bigcode project released [StarCoder](https://huggingface.co/bigcode/starcoderbase). Technically the plugin can support any model, you either need to supply a model ID (this will work with our API inference service at Hugging Face) or an http endpoint.

It may be worth joining forces, interested in collaborating?

**Comment by jukefr** [comment]
Tried out https://github.com/fauxpilot/fauxpilot really quickly (and was not all that impressed tbh but i can only reliably run the 350M parameter model)
but still, made a quick fork of this with an action that auto merges upstream daily at like 1am and `sed` patches the agent.js to use the local fauxpilot install if people are interested 🤷 https://github.com/jukefr/fauxpilot.lua

<details>
Would probably be neat to be able to configure this somehow instead of having to do this botch but from what i gather this project just syncs the `dist` folder from official copilot plugin anyway so it would imply a rewrite/rethink of some sort and this is quick and dirty but works 🤷
</details>
